File: 239.sliding-window-maximum.java
/*
 * @lc app=leetcode id=239 lang=java
 *
 * [239] Sliding Window Maximum
 */

// @lc code=start
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        int[] maxSWindow = new int[nums.length - k + 1];
        int left = 0, right = k - 1;
        int maxIdx = getMaxIdx(nums, 0, k - 1);

        while (right < nums.length) {
            if (left <= maxIdx) {
                // max is in the window
                maxSWindow[left] = nums[maxIdx];
                left++;
                right++;

                // prevent java.lang.ArrayIndexOutOfBoundsException
                // from if (nums[right] >= nums[maxIdx])
                if (right == nums.length)
                    break;

                if (nums[right] >= nums[maxIdx])
                    maxIdx = right;
            } else {
                // max is out of the window

                // This 2 if-else make it form 500ms to 3ms !!!
                if (nums[right] >= nums[maxIdx] - 1)
                    maxIdx = right;
                else if (nums[left] >= nums[maxIdx] - 1)
                    maxIdx = left;
                else
                    maxIdx = getMaxIdx(nums, left, right);
            }
        }
        return maxSWindow;
    }

    private int getMaxIdx(int[] nums, int left, int right) {
        int maxVal = nums[left], maxIdx = left;
        for (int i = left + 1; i <= right; i++)
            if (nums[i] >= maxVal) {
                maxVal = nums[i];
                maxIdx = i;
            }
        return maxIdx;
    }

}
// @lc code=end

File: 239.sliding-window-maximum.py
#
# @lc app=leetcode id=239 lang=python3
#
# [239] Sliding Window Maximum
#


# @lc code=start
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        n = len(nums)
        # # stores pairs (index, value)
        # in val's decreasing value order
        dq = deque()
        res = []

        # Initialize deque with the first window
        for i in range(k):
            # Remove all smaller values from the back,
            # since they cannot be the maximum if a larger arrives
            while dq and nums[i] >= dq[-1][1]:
                dq.pop()
            dq.append((i, nums[i]))
        res.append(dq[0][1])

        # Process the rest of the windows
        for i in range(k, n):
            # Remove the element that slid out of the window
            if dq and dq[0][0] == i - k:
                dq.popleft()

            while dq and nums[i] >= dq[-1][1]:
                dq.pop()
            dq.append((i, nums[i]))
            res.append(dq[0][1])

        return res


# @lc code=end
